CDAP-5878 : Added Documentation for Value Mapper Plugin and Three field input Widget

ameyk opened the pull request:
@sreevatsanraman @albertshau @anwar6953 @ajainarayanan 
Added documentation for value mapper plugin. Added three field input widget for mapping input field.

Commit:
Correct description of transform.

Commit:
Correct spelling and add a deprecation notice.

Commit:
"JavaScript" Typos.

Commit:
Modified widget json to add three field widget for mapping input parameter

Commit:
Modified widget json to add three field widget for mapping input parameter

Commit:
Added documentation file for Value Mapper plug-in

Commit:
Added documentation file for Value Mapper plug-in

Commit:
Added missing bracket

Commit:
Added missing bracket

anwar6953 commented on transform-plugins/widgets/ValueMapper-transform.json:
```diff
@@ -5,12 +5,18 @@
+	  "label": "mapping",
+	  "widget-type": "ds-multiplevalues",
+	  "widget-attributes": {
+	  "delimiter": ",",
```
should indent lines 13-17, since they're a part of the 'widget-attributes' field

anwar6953 commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+
+Use Case
+--------
+Suppose, If you want to replace language codes in the input record field to language description.
```
don't capitalize If

awholegunch replied:
```
One use is to replace language codes in the input record field with 
an appropriate language description:

     **Source field name:** language_code
     **Target field name:** language_desc
     **Mappings, source to target:** DE/German, ES/Spanish, EN/English

This will replace the source column *language_code* with the target column 
*language_desc*, replacing values found in the source field using the mappings 
("DE" to "German", "ES" to "Spanish", and so on.)
```

anwar6953 commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+Description
+-----------
+Value Mapper is a transform plugin that maps string values of a field in the input record to mapping value using mapping dataset.
+Mappings are usually stored in KeyValue dataset. Value Mapper provides you a simple alternative for manipulating the input data by using mapping dataset
```
add period at the end

awholegunch replied:
```
Mappings for the values are usually stored in a key-value dataset. The ValueMapper transform
provides a simple method for manipulating input data, both a field and its values, using a mapping.
```

anwar6953 commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+
+Properties
+----------
+**mapping:** Defines mapping of source field to target field and mapping table name for the specified source field. 
```
Is it possible to map more than one field?

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+
+Description
+-----------
+Value Mapper is a transform plugin that maps string values of a field in the input record to mapping value using mapping dataset.
```
Please make the lines shorter (max 120, though 100 is better) as then GitHub doesn't wrap them when we review...

```
ValueMapper is a transform plugin that maps field names and the string values of a field in
the input record to mapping values using a mapping dataset. Both the field name and its 
values are mapped.
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+
+**defaults:** Contains a key value pair of source field and its default value in case if the source field value is
+NULL/EMPTY or if the mapping value is not present. If no default value has been provided, then the source field value will be mapped to target field. 
+Value accepted is of STRING NULLABLE type only.
```
```
**mapping:** A comma-separated list that defines the mapping of a source
field to a target field and the mapping table name for looking up values. 
Contains three fields separated by a colon (":") as the source field, the 
mapping table name, and the target field:

         <source-field>:<mapping-table-name>:<target-field>

Note: **source field** supports only STRING types.

**defaults:** A comma-separated list that contains key-value pairs of a
source field and its default value for cases where the source field
value is either null or empty or if the mapping value is not present. If
a default value has not been provided, the source field value will be
mapped to the target field. Only STRING NULLABLE type values are accepted.
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+Example
+-------
+Suppose user takes the input data(employee details) through stream and wants to apply
+value mapper transform plugin on designation field
```
```
As an example, take employee details as input data through a stream and then apply
the ValueMapper transform on the *designation* field in the input data.
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+Suppose user takes the input data(employee details) through stream and wants to apply
+value mapper transform plugin on designation field
+
+Plugin JSON Representation will be:
```
The plugin JSON representation will be:

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+    }
+
+
+If the transform receives following input record:
```
If the transform receives as an input record:

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+    | designation| string              | "2"                  |
+    +=========================================================+
+
+Sample structure for Mapping/Lookup Dataset
```
with this as the mapping dataset:

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,82 @@
+    +=======================+
+
+After the transformations from ValueMapper plugin, output will have below structure.
+Here destination column values will be replaced in destinationName by using Mapping Dataset (Key-Value pair):
```
```
After transformation by the ValueMapper plugin, the output will have this structure and contents, with the
*designation* column replaced by the *designationName* column, using values looked up from the 
mapping database:
```

awholegunch commented on transform-plugins/widgets/ValueMapper-transform.json:
```diff
@@ -5,12 +5,18 @@
-        },
-        {
+       {
+       	  "name": "mapping",
```
This should be de-dented so that "name" lines up with "label".

awholegunch commented on transform-plugins/widgets/ValueMapper-transform.json:
```diff
@@ -5,12 +5,18 @@
+	  "delimiter": ",",
+	  "values-delimiter": ":",
+	  "numValues": "3",
+	  "placeholders": ["Source field", "Lookup table name", "Target field"]
```
```
"placeholders": ["Source Field", "Mapping Table Name", "Target Field"]
```

awholegunch commented:
I've reviewed this, and can take another look after it has been revised.

Commit:
Initial version of parsing complex json

Commit:
Adding in widget for mapping

Commit:
Adding documentation for plugin

Commit:
Adding documentation for plugin

Commit:
Fix Json-xpath

cdap-apps committed:
Added capability to copy input field values into output fields if there are no mappings specified

cdap-apps committed:
Fixing checkstyle issues

cdap-apps committed:
Fixing comment for removing spaces in MD tables

cdap-apps committed:
CDAP-6097 Fixing issue with checking with case insensitivity on
Added ability to pass through input record fields into output if field names are the same

Commit:
Merge branch 'develop' into feature/doc_fixes

Commit:
Removes the "description" fields as they are redundant. This information is obtained from the back-end.

romy-khetan commented:
Updated document and widget files as per above comments and created a new pull request #273 with the modified files.

Commit:
Updated documentation as per review comments

Commit:
Updated documentation as per review comments

Commit:
Updated widget as per review comments

Commit:
Updated widget as per review comments

elfenheart committed:
add placeholders to widgets JSONs

rashig committed:
JIRA ID - CDAP-5880.Implement Row Denormalizer Plugin for Hydrator. Initial Draft (#259)

* JIRA ID - CDAP-5880.Implement Row Denormalizer Plugin for Hydrator.Initial Draft

* JIRA- CDAP-5880. Review comment implementation.

* JIRA- CDAP-5880. Review comment implementation.

* JIRA- CDAP-5880. Review comment implementation.

* JIRA- CDAP-5880. Review comment implementation.

* JIRA- CDAP-5880. Review comment implementation.

* JIRA- CDAP-5880. Review comments implementation.

prinam commented:
#273 has been closed and this PR has been updated with latest changes. 

@ajainarayanan @awholegunch please take a look.

awholegunch commented on transform-plugins/widgets/ValueMapper-transform.json:
```diff
@@ -5,10 +5,16 @@
       "properties": [
-        {
-          "widget-type": "textbox",
+	    {
```
This is not lined up correctly with the following (closing) brace. It should be moved to the left.

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+Description
+-----------
+Value Mapper is a transform plugin that maps string values of a field in the input record 
+to mapping value using mapping dataset.
```
to a mapping value using a mapping dataset.

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+
+This will replace the source column *language_code* with the target column 
+*language_desc*, replacing values found in the source field using the mappings 
+("DE" to "German", "ES" to "Spanish", and so on.)
```
Remove parentheses:

```
*language_desc*, replacing values found in the source field using the mappings 
"DE" to "German", "ES" to "Spanish", and so on.
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+----------
+**mapping:** A comma-separated list that defines the mapping of a source
+field to a target field and the mapping table name for looking up values. 
+Contains three fields separated by a colon (":") as the source field, the 
```
Contains three properties

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+
+**defaults:** A comma-separated list that contains key-value pairs of a
+source field and its default value for cases where the source field
+value is either null or empty or if the mapping value is not present. If
```
the mapping key-value

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+    | designation| string              | "2"                  |
+    +=========================================================+
+
+with this as the mapping dataset:
```
```
with this as the mapping dataset, *designationLookupTableName*:
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+        "type": "transform",
+        "properties": {
+            "mapping": "designation:designationLookupTableName:designationName",
+            "defaults": "designation:DefaultDesignation"
```
How will `DefaultDesignation` be used? Where is it defined? Can you add an example showing this being used?
Is it to be one of the rows of the key-value table?

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+    | 1          | SE       |
+    | 2          | SSE      |
+    | 3          | ML       |
+    +=======================+
```
Should these rows be shown as Strings?

```
    +=======================+
    | key        | value    |
    +=======================+
    | "1"        | "SE"     |
    | "2"        | "SSE"    |
    | "3"        | "ML"     |
    +=======================+
```

awholegunch commented on transform-plugins/docs/ValueMapper-transform.md:
```diff
@@ -0,0 +1,99 @@
+
+         <source-field>:<mapping-table-name>:<target-field>
+
+Note: **source field** supports only STRING types.
```
Given that source field only supports STRING types, does that mean that the keys all have to be strings, though the mapped values could be a different type, if the column it was mapped to (the target field) supported it?

elfenheart commented on transform-plugins/widgets/ValueMapper-transform.json:
```diff
@@ -5,10 +5,16 @@
-          "widget-type": "textbox",
+	    {
+          "name": "mapping",
           "label": "mapping",
```
For display purpose, we should have the label to start with Uppercase letter. This applies to line 21 as well

nitinmotgi committed:
Merge pull request #276 from caskdata/feature/add-placeholder-widget-options

[CDAP-5667] add placeholders to widgets JSONs

ameykulk committed:
Incorporated review comments

ameykulk committed:
Incorporated review comments

ameykulk committed:
Incorporated review comments

ameykulk committed:
Incorporated review comments

Commit:
Merge pull request #272 from caskdata/feature/doc_fixes

Documentation fixes

gokulavasan committed:
CDAP-6140 binary is a reserved keyword. hence use backticks in explore

Using backticks for binary since it is a keyword

gokulavasan committed:
Merge pull request #278 from caskdata/fix/travis-build-failure

CDAP-6140: binary field name seems to cause trouble with explore

gokulavasan committed:
Merge pull request #271 from caskdata/feature/csv-parser-2

Added ability to pass through input record fields into output if field

ananya-bhattacharya committed:
JIRA ID- CDAP-5879. Implement COBOL Copybook Plugin for Hydrator.(Initial draft) (#258)

* JIRA ID- CDAP-5879. Implement Cobol Copybook Plugin for Hydrator.(Initial draft)

* jar files required for JRecord classes are used from the local repo folder

* JIRA ID- CDAP-5879.Code Review Comment Implementation.

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Implemented test case for switch statement.

* Removed unused constants

cdap-apps committed:
Addressing PR comments

cdap-apps committed:
Addressing PR comments

cdap-apps committed:
Merge branches 'feature/json-xpath' and 'develop' of github.com:caskdata/hydrator-plugins into develop

cdap-apps committed:
Row key inclusion in output schema is optional. Change it from being mandatory.

nitinmotgi committed:
Update JSONParser-transform.md

Addressed PR comments.

nitinmotgi committed:
Merge pull request #270 from caskdata/feature/json-xpath

Support for Parsing complex Json using Json Path

cdap-apps committed:
Merge branch 'develop' of github.com:caskdata/hydrator-plugins into develop

cdap-apps committed:
Merge branch 'develop' into feature/table-no-key-outputschema

nitinmotgi commented:
LGTM 👍 . Merge it please.

prinam commented:
@ameyk looks like there are some merge conflicts. once those are resolved, this PR can be merged.

gokulavasan committed:
Fix develop build failure by creating a common test base for transfor… (#286)

* Fix develop build failure by creating a common test base for transform module and add all the plugins used to the plugin artifact

* Ignore copybook tests since CDAP-6162 is blocking it

nitinmotgi committed:
Merge pull request #282 from caskdata/feature/table-no-key-outputschema

Table Sink : Row key inclusion in output schema is optional now

ananya-bhattacharya committed:
CopybookReader - CDAP 6162, fixed (#287)

* JIRA ID- CDAP-5879. Implement Cobol Copybook Plugin for Hydrator.(Initial draft)

* jar files required for JRecord classes are used from the local repo folder

* JIRA ID- CDAP-5879.Code Review Comment Implementation.

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Code review comment implementation

* Implemented test case for switch statement.

* Removed unused constants

* CopybookReader bug - CDAP-6162, fixed

Commit:
Removed valuemapper json file to resolve conflicts

Commit:
Removed valuemapper json file to resolve conflicts

Commit:
Merge branch 'develop' of https://github.com/ameyk/hydrator-plugins into develop
merging changes of forked branch

prinam commented:
Closing this PR as there are unnecessary files in this one. A new one with the right file will be opened.
